package/MDAnalysis/lib/_augment.pyx

```diff
@@ -0,0 +1,353 @@
+#
+# Please cite your use of MDAnalysis in published work:
+#
+# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+#
+
+import cython
+import numpy as np
+from .mdamath import triclinic_vectors
+
+from libcpp.vector cimport vector
+from libc.math cimport sqrt
+
+__all__ = ['augment_coordinates', 'undo_augment']
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def augment_coordinates(float[:, ::1] coordinates, float[:] box, float r):
+    """Calculates the relevant images of particles which are within a
+    distance 'r' from the box walls
+
+    Only the relevant images are generated i.e. for every set of
+    coordinates close to the box boundary,
+    corresponding images which are close to the opposite face and
+    outside the central cell are generated. If the particle
+    is close to more than one box walls,
+    images along the diagonals are also generated ::
+
+
+                       |  x               x
+    +---------------+  |    +---------------+
+    |               |  |    |               |
+    |               |  |    |               |
+    |               |  |    |               |
+    |             o |  |  x |             o |
+    +---------------+  |    +---------------+
+                       |
+
+    Parameters
+    ----------
+    coordinates : array
+      Input coordinate array to generate duplicate images
+      in the vicinity of the central cell. All the coordinates
+      must be within the primary unit cell.
+    box : array
+      Box dimension of shape (6, ). The dimensions must be
+      provided in the same format as returned
+      by :attr:`MDAnalysis.coordinates.base.Timestep.dimensions`:
+      ``[lx, ly, lz, alpha, beta, gamma]``
+    r : float
+      thickness of cutoff region for duplicate image generation
+
+    Returns
+    -------
+    output : array
+      coordinates of duplicate(augmented) particles
+    indices : array
+      original indices of the augmented coordinates
+      A map which translates the indices of augmented particles
+      to their original particle index such that
+      ``indices[augmentedindex] = originalindex``
+
+    Note
+    ----
+    Output doesnot return coordinates from the initial array.
+    Use `np.concatenate(coordinates, output)` to merge particle
+    coordinates as well as their images.
+
+    .. SeeAlso:: :func:'MDAnalysis.lib._augment.undo_augment'
+

```

Comment:
The Notes should also describe the algorithm that is used. 

We might want to describe it in an upcoming MDAnalysis paper but in any case, it looks sufficiently complicated so that just reading the code below does not immediately convey the key ideas. 

Does the algorithm provably find all images or can it miss something?

Reply:
1. The Tuckerman algorithm does *NOT* work for triclinic boxes. It assumes the box vectors are perpendicular. The explicit algorithm is your only chance if you work without a cutoff. If your cutoff is below half the box-length a faster version exists (The old implementation in MDAnalysis pre 0.17.0).
2. If you project along the box vectors you will find all images. There are 26 combinations to use to get all images. A projection algorithm will work independently of box type.

Reply:
@orbeckst I don't think the algorithm is actually that fancy, it's just long.  For a 1D case it just finds particles close to the edges and creates a new image a box length away.  For the 2D case you find particles close to either of the edges, and moves them a box length, then particles which are close to both edges get also moved diagonally (this is illustrated in the ascii art).  I think should find all particles.

@ayushsuhane for single queries yes, but we rarely calculate a single pairwise distance.  For the all vs all case we have for guess_bonds this explicit image solution works pretty well with KDT.

Reply:
**Caution, mild rant**

Properly writing down a new algorithm (equations, pseudocode) is important:
1. We should strive to communicate what's been done in a manner suitable for anyone else to re-implement it – they shouldn't need to look at your code for that. And just because the algorithm is "trivial" to the creator (at this instance in time... what about 6 months later?) it really is seldom trivial to the vast majority of people who will spend time looking at it or reading about it. 
2. Writing things down clearly is a good way to find mistakes and to give other people a chance to find something that you might have overlooked – we all have our own biases and we should welcome any opportunity to have our thinking checked.
3. If the algorithm is not new then there needs to be a proper citation (if available) and an attribution to the source. (But being clear in our docs and code is still important for maintenance – by your future self and the developers around and after you).

Lastly, being clear is just doing the scholarly right thing.